Added the New Configuration Manager invoked through a command line arg that allows the user to create new configurations based on a prompt

MokoSan opened the pull request:
By passing in ``-g <Path to config>``, the user is able is create and persist a new config file based on a prompt such as the following:

![C__Users_mukun_source_repos_realmon_src_GCRealTimeMon_bin_Debug_net6 0_GCRealTimeMon exe 2021-11-19 22-21-57](https://user-images.githubusercontent.com/4951960/142716885-816f6ad5-bc2c-4307-b49c-4f56cf36fe74.gif)

Based on the choices of the prompt and the path provided in the command line, a new config file will be persisted and can be used in subsequent runs. The choice of command line args are:

If neither -g nor -c is specified, deserialize: DefaultConfig.yaml.
If -c is specified without an argument, open up the prompt, overwrite DefaultConfig.yaml and make use of the the new configs in the session. <- this was a bit hacky considering the CommandLineParser Options are strongly typed i.e. I couldn't specify both a bool and a string for the path as the type of that argument.
If -c is specified with an argument, deserialize the path specified as the argument.
If -g is specified, open up the prompt, persist the new settings in the path specified and make use of the new configs in the session.

MokoSan committed:
Added a command line param that allows for the custom creation of the config via prompts.

MokoSan committed:
Updated the README with the new command line args

MokoSan committed:
Fixed merge conflicts -- need to resolve some issues

chrisnas commented:
> By passing in `-g <Path to config>`, the user is able is create and persist a new config file based on a prompt such as the following:
> 
> ![C__Users_mukun_source_repos_realmon_src_GCRealTimeMon_bin_Debug_net6 0_GCRealTimeMon exe 2021-11-19 22-21-57](https://user-images.githubusercontent.com/4951960/142716885-816f6ad5-bc2c-4307-b49c-4f56cf36fe74.gif)
> 
> Based on the choices of the prompt and the path provided in the command line, a new config file will be persisted and can be used in subsequent runs.

This is a great idea to allow the user to select the columns in the console!
However, why not simply using -c without parameter to indicate that columns will be selected and saved in the default .yaml file? That way, not need to deal with new config file  :^)

In addition, you should rebase with the current state of the code because it has been refactored a lot; especially for parsing parameters and handle the session in a more consistent way.

Last but not least, don't forget to update the version in the dotnet-gcmon.csproj file

MokoSan committed:
Re-added the New Configuration creator

MokoSan committed:
Added the ability to not specify an argument if -c is passed and defaults the behavior to overwrite the default config

MokoSan committed:
Bumped up the version of the dotnet-gcmon project and added all the new files

MokoSan commented:
> This is a great idea to allow the user to select the columns in the console! However, why not simply using -c without parameter to indicate that columns will be selected and saved in the default .yaml file? That way, not need to deal with new config file :^)

Thanks for the great idea!! I have included the following behavior:

1. If neither -g nor -c is specified, deserialize: DefaultConfig.yaml.
2. If -c is specified without an argument, open up the prompt, overwrite DefaultConfig.yaml and make use of the the new configs in the session. <- this was a bit hacky considering the CommandLineParser Options are strongly typed i.e. I couldn't specify both a ``bool`` and a ``string`` for the path as the type of that argument. Either way, happy I got this working. 
3. If -c is specified with an argument, deserialize the path specified as the argument.
4. If -g is specified, open up the prompt, persist the new settings in the path specified and make use of the new configs in the session.

> 
> In addition, you should rebase with the current state of the code because it has been refactored a lot; especially for parsing parameters and handle the session in a more consistent way.
> 

Done! Totally forgot to rebase but now have fixed all merge conflicts. Thanks for catching this!

> Last but not least, don't forget to update the version in the dotnet-gcmon.csproj file
Done!

Thank you for all the great feedback, @chrisnas! Please keep it coming if you think anything else can be improved.

Maoni0 commented:
thank you both!! :)

if I think about how I would use a new tool, I'd probably start by typing "dotnet gcmon /?" or simply "dotnet gcmon" just to see what it tells me. right now it says `Specify a process Id using: -p or a process name by using -n.`. what if we displayed the help when the user doesn't specify an arg or specifies "/?"? like this  

```
Usage -
Specify a process Id by using -p or a process name by using -n, the tool will show GCs as they occur in that process. If there are multiple processes with that name it would pick the first one
Specify a min GC pause in ms by using "-m x" to only display GCs with pauses above x ms
You can specify which info to display per GC by using a config file. You can either change the current config at [fill in the path to DefaultConfig.yaml here] or specify your own by using "-c config_file_path", eg, "-c c:\data\gcmon-config.yaml" 
To create a config file, use -g which allows you to specify a path for the config file and choose which info to display for each GC
```

does this seem like it would make starting using this tool easier? (please feel free to make the text more clearl)

also just a side thought, since we have a config file, we should probably deprecate the "-m" option and make that part of the config?

MokoSan committed:
Removed the -m command line arg and replaced it with an optional config & added help text

MokoSan commented:
@Maoni0, I think your incorporated feedback definitely improves the overall user experience. :)

I have added the following help text in case the user:
- doesn't specify _any_ command line args.
- specifies "?" in the command line or "-- /?".
- the process id / process name aren't specified.

The final help text looks like the following:
![image](https://user-images.githubusercontent.com/4951960/142751572-683cbea7-6c37-48da-9b89-4581ac8a716e.png)

I have also removed the -m command line arg and replaced it with specifying the following in the config:

```
display_conditions:
    min gc duration (msec): 200
```

Here is a gif of the new prompt menu:

![C__Users_mukun_source_repos_realmon_src_GCRealTimeMon_bin_Debug_net6 0_GCRealTimeMon exe 2021-11-20 22-04-52](https://user-images.githubusercontent.com/4951960/142751703-420177be-c628-4544-9d4e-b01bdba19a57.gif)

Maoni0 commented:
thanks! I just tried this out. BTW I should clarify what I had here in the help text -

> You can specify which info to display per GC by using a config file. You can either change the current config at [fill in the path to DefaultConfig.yaml here] or specify your own by using "-c config_file_path", eg, "-c c:\data\gcmon-config.yaml"

I meant `[fill in the path to DefaultConfig.yaml here]` is to be replaced by the actual path of DefaultConfig.yaml on the machine :)

> To create a config file, use -g which allows you to specify a path for the config file and choose which info to display for each GC or overwrite the default config by entering -c without any parameters.

I tried `GCRealTimeMon.exe -c` and `GCRealTimeMon.exe -g c:\temp\myconfig.yaml`but it seems to tell me the same help text... from the text it sounds like `GCRealTimeMon.exe -c` should also bring up the part that allows me to select which info to display except it should be pre-populated with what's in the default?

MokoSan committed:
Added the ability to read from defaults and updated the help text + way the configs are picked up

MokoSan commented:
> I meant `[fill in the path to DefaultConfig.yaml here]` is to be replaced by the actual path of DefaultConfig.yaml on the machine :)

Fixed.. lol

> I tried GCRealTimeMon.exe -c and GCRealTimeMon.exe -g c:\temp\myconfig.yamlbut it seems to tell me the same help text... from the text it sounds like GCRealTimeMon.exe -c should also bring up the part that allows me to select which info to display except it should be pre-populated with what's in the default?

I should have updated this logic.. What I had before was that we absolutely needed the pid / process name for anything to happen. With the new logic, if the user passes ``-c`` or ``-g <path>``, they will automatically go to the configuration creation mode. 

Now, for the ``-c`` command line arg, I am reading the selected values e.g. columns from the default. For example, for the following config:

```
columns:
- type
- pause (ms)
- gen0 alloc (mb)
stats_mode:
  timer: 20s
display_conditions:
  min gc duration (msec): 200
```

The prompts will be:
![C__Users_mukun_source_repos_realmon_src_GCRealTimeMon_bin_Debug_net6 0_GCRealTimeMon exe 2021-11-20 23-51-31](https://user-images.githubusercontent.com/4951960/142754229-ccc68656-464e-492c-abd6-ea5e7906dce5.gif)

For the sake of completeness, here are the main actions as per the user specified command line args:

| Command Line               | Action                                                                                                    |
|----------------------------|-----------------------------------------------------------------------------------------------------------|
| No Args                    | Help displayed                                                                                            |
| -? or --/?                 | Help Displayed                                                                                            |
| -c                         | Prompt to generate a config file to overwrite the DefaultConfig.yaml with defaults from that file itself. |
| -g (path)                  | Prompt to generate a config file in the desired path.                                                     |
| -c -n (processName)        | Same as -c but this time, realmon will also start the monitoring.                                         |
| -g (path) -n (processName) | Same as -g (path) but this time, realmon will also start the monitoring.                                  |
| -n (processName)           | Configs are read from DefaultConfig.yaml and realmon will start the monitoring.                           |

MokoSan committed:
Fix for null refs.

chrisnas commented:
> For the sake of completeness, here are the main actions as per the user specified command line args:
> 
> Command Line	Action
> No Args	Help displayed
> -? or --/?	Help Displayed
> -c	Prompt to generate a config file to overwrite the DefaultConfig.yaml with defaults from that file itself.
> -g (path)	Prompt to generate a config file in the desired path.
> -c -n (processName)	Same as -c but this time, realmon will also start the monitoring.
> -g (path) -n (processName)	Same as -g (path) but this time, realmon will also start the monitoring.
> -n (processName)	Configs are read from DefaultConfig.yaml and realmon will start the monitoring.

Here is what you get for other dotnet CLI tools such as dotnet counters which is close to what gcmon is doing:
![image](https://user-images.githubusercontent.com/10163781/142757061-0966ab02-f962-41c9-a80f-b81f0e2e1aea.png)

So help is displayed like you implemented (no parameter, `-?`) but also with `-h` and `--help`.
If a command is given without mandatory options, the corresponding help is provided.
And `-h` for a given command details the options for that specific command:
![image](https://user-images.githubusercontent.com/10163781/142757135-f2c5524a-e9da-4d05-a4e2-d3157999061c.png)

Also, each command is exclusive and options apply to commands.
In our case, we could have the following commands:
`   monitor -p <pid> -c <input yaml file - if not provided, enter your great column selection mode>`
`   configure <output yaml file - if not provided, enter your great column selection mode and save into default yaml>`

Also, we could imagine supporting the same `collect` command that would save the result in a .csv file. Since I'm thinking about the ways to use **dotnet gcmon** in scripts, the `-d(uration)` option for `collect` command could be interesting so it would run for the given duration and save the output into the given file:
`dotnet gcmon collect -p <pid> -n <name> -o(utput) <file name> -f(ormat) <csv|json - csv is the default if not provided>`
I'm willing to implement that new `collect` command in another PR so you don't add unrelated code to your PR if you prefer :^)

MokoSan commented:
Great suggestions - will clean up the help text unless there is a way to automatically do it. The collect command seems like a great addition.

MokoSan committed:
Improved the usage text

MokoSan commented:
@chrisnas @Maoni0:

Below what the new usage text looks like, please let me know if any other changes are needed. This shows up iff:
- The user enters _no_ command line args i.e. ``GCRealTimeMon.exe``.
- The user asks for help i.e. ``GCRealTimeMon.exe -?`` or ``GCRealTimeMon.exe --\?``.
- The user doesn't provide the necessary args i.e. either the process name or the process id:

![image](https://user-images.githubusercontent.com/4951960/142803105-892e2a5c-6a7e-4965-8c23-776d2a9cbc8b.png)

MokoSan committed:
Fix for LOH based naming.

Maoni0 commented on src/GCRealTimeMon/Utilities/CommandLineUtilities.cs:
```diff
@@ -0,0 +1,64 @@
+
+More Details:
+    - Specify a process Id by using -p or a process name by using -n, the tool will show GCs as they occur in that process. If there are multiple processes with that name it would pick the first one
+    - You can specify which info to display per GC by using a config file. You can either change the current config at ``GCRealTimeMon\bin\Release\net6.0\DefaultConfig.yaml`` or specify your own by using ""-c config_file_path"", eg, ""-c c:\data\gcmon-config.yaml"" 
```
>GCRealTimeMon\bin\Release\net6.0\DefaultConfig.yaml` [](http://example.com/codeflow?start=121&length=52)

this would need to be obtained programmatically, please see [here](https://github.com/Maoni0/realmon/blob/main/src/GCRealTimeMon/Program.cs#L183). now that I think more about it, it would need to say "if you are using this as a dotnet global tool it's the path where gcmon dll is, but if you are using this when you build it yourself, it's the DefaultConfig.yaml where GCRealTimeMon.dll is

MokoSan committed:
updated the help text to programmatically insert the location of the default path

Maoni0 commented:
when I answered 'n' to 

`Would you like to set a value for the minimum GC Pause duration to filter GCs off of?`

it just shows me the help text again

![image](https://user-images.githubusercontent.com/10837357/142818161-a1b9beb6-14c6-4eb6-ab5f-27857d2295aa.png)

also the selected columns are very difficult to read being in one very long line.. would be great to keep each column as its own line. 

I'm unclear what this means -

```
  -c, --configPath          The path to the YAML configuration file that is read
                            in. If no path is specified, the user can create a
                            new config via prompts on the command line.
```

when I do -c with no path, it's supposed to populate the options from my current default config, correct? it doesn't look that way... 

for -g

```
  -g, --createConfigPath    The path of the config to be created via the command
                            line prompt.
``` 

I also got the same result - it just shows me the help text again after I answer a question -

![image](https://user-images.githubusercontent.com/10837357/142819088-c54340a1-72e5-447c-a9b2-3e75d028ba44.png)

but it did create c:\temp\config.yaml.

I also noticed in the .yaml there's no longer comments in available_columns. it would be useful to still have these when the user edits the config by hand.

MokoSan committed:
Addressed all the comments/feedback re: the user experience of the command line + configs.

chrisnas commented on README.md:
```diff
@@ -9,8 +9,8 @@ Right now it's super simple - given a PID or process name it will show you a few
+| c | Path of the Configuration File | Path to the configuration file used by the monitor. This file is a YAML file. By default, it's the Default.yaml file is loaded. If no argument is specified, the command line prompt to create the config will be displayed and the default config will be overwritten with the inputted values; this new config will be used for this session. | 
+| g | Path of the New Configuration File | Path to the configuration file that'll be created using a command line prompt, persisted to the given path and used by the monitor for this particular session. This file will be of the YAML format. |
 
 Note: Either the name of the process or the process id must be specified, else an ``ArgumentException`` is thrown.
```
This should not be the case: argument exceptions are handled and proper error message is provided.
Also, should not be mandatory when -g or -c

chrisnas commented on README.md:
```diff
@@ -9,8 +9,8 @@ Right now it's super simple - given a PID or process name it will show you a few
 | p | Process Id | Process Id to monitor GC for |
-| m | Minimum Duration in Ms for GC Pause Duration | Any GC Pauses below this specified value will not be written to the console log - this is an optional argument |
-| c | Path of the Configuration File | Path to the configuration file used by the monitor. This file is a YAML file. By default, it's the Default.yaml file is loaded | 
+| c | Path of the Configuration File | Path to the configuration file used by the monitor. This file is a YAML file. By default, it's the Default.yaml file is loaded. If no argument is specified, the command line prompt to create the config will be displayed and the default config will be overwritten with the inputted values; this new config will be used for this session. | 
```
```suggestion
| c | Path of a YAML configuration File to use | This can be used to override default configuration either through another file or manual column selection that will also update the default when no file is provided. |
```

chrisnas commented on README.md:
```diff
@@ -9,8 +9,8 @@ Right now it's super simple - given a PID or process name it will show you a few
-| m | Minimum Duration in Ms for GC Pause Duration | Any GC Pauses below this specified value will not be written to the console log - this is an optional argument |
-| c | Path of the Configuration File | Path to the configuration file used by the monitor. This file is a YAML file. By default, it's the Default.yaml file is loaded | 
+| c | Path of the Configuration File | Path to the configuration file used by the monitor. This file is a YAML file. By default, it's the Default.yaml file is loaded. If no argument is specified, the command line prompt to create the config will be displayed and the default config will be overwritten with the inputted values; this new config will be used for this session. | 
+| g | Path of the New Configuration File | Path to the configuration file that'll be created using a command line prompt, persisted to the given path and used by the monitor for this particular session. This file will be of the YAML format. |
```
```suggestion
| g | Path of a YAML configuration file to generate | Select the columns to display and the corresponding configuration will be saved to the given YAML file and used for the current monitoring session.|
```

chrisnas commented on src/GCRealTimeMon/Configuration/ConfigurationReader.cs:
```diff
@@ -18,12 +20,17 @@ internal static class ConfigurationReader
         public static async Task<Configuration> ReadConfigurationAsync(string path)
         {
-            IDeserializer deserialier = new DeserializerBuilder()
+            if (string.IsNullOrEmpty(path))
```
The path can never be null + ArgumentNullException is not properly handled in calling code

chrisnas commented on src/GCRealTimeMon/Configuration/NewConfigurationManager.cs:
```diff
@@ -0,0 +1,138 @@
+                                 .WithNamingConvention(UnderscoredNamingConvention.Instance)
+                                 .Build();
+            var serializedResult = serializer.Serialize(configuration);
+            await File.WriteAllTextAsync(path: path, contents: serializedResult);
```
Is there any chance that path could be null or invalid?
Maybe at least protect against the invalid situation with a try/catch and transform into:
`   throw new ArgumentException($"Invalid configuration file path '{path}'");`

MokoSan committed:
Fixed the defaults read in from the configuration

MokoSan commented:
> it just shows me the help text again

I have removed the help menu display if the user passes in:
``-c`` without the path OR ``-g <path>`` _BUT_ no process id / process name. This does seem like the more user friendly choice. My previous rationale was that the process id / process name _must_ be passed always or else, we'd display the usage via the help text (was going off @chrisnas' example).

> also the selected columns are very difficult to read being in one very long line.. would be great to keep each column as its own line.

This has been fixed. Example:

__Prompt__:
![image](https://user-images.githubusercontent.com/4951960/142833722-27b0ade0-c01b-4256-ad4b-33ae318aaa13.png)

__Post-Selection__:
![image](https://user-images.githubusercontent.com/4951960/142833554-02135c24-76c1-4eff-a172-edb93b71c85e.png)



> I'm unclear what this means -
> 
> ```
>   -c, --configPath          The path to the YAML configuration file that is read
>                             in. If no path is specified, the user can create a
>                             new config via prompts on the command line.
> ```

I think I have improved the description here but would like a second opinion / any further improvements.

```
  -c, --configPath          The path to the YAML configuration file that is used
                            in the session; this configuration file contains
                            information such as columns to display that can be
                            configured accordingly. If no path is specified, the
                            default configuration is overwritten by the answers
                            to the prompts in the command line.
```


> when I do -c with no path, it's supposed to populate the options from my current default config, correct? it doesn't look that way...

I think I fixed this. How I intended this to work is, the details from the defaults such as in the case of columns:

```
columns:
- gen # The Generation
- suspension time (ms) # The time in milliseconds that it took to suspend all threads to start this GC 
- gen0 alloc (mb) # Amount allocated in Gen0 since the last GC occurred in MB.
- gen0 alloc rate  #  The average allocation rate since the last GC.
```

the prompt would be pre-populated as:
![image](https://user-images.githubusercontent.com/4951960/142837315-2c1e2d1e-dec6-413e-a988-669aef876fb2.png)

> I also noticed in the .yaml there's no longer comments in available_columns. it would be useful to still have these when the user edits the config by hand.

Have programmatically added as many comments as possible. Example:

```
# Selected Columns to Display
columns:
- gen # The Generation
- suspension time (ms) # The time in milliseconds that it took to suspend all threads to start this GC 
- gen0 alloc (mb) # Amount allocated in Gen0 since the last GC occurred in MB.
- gen0 alloc rate #  The average allocation rate since the last GC.
# All the Columns Available to Choose From. Please refer to the table via the following link for details about the columns: https://github.com/Maoni0/realmon#configuration
available_columns:
- index # The GC Index.
- type # The Type of GC.
- gen # The Generation
- pause (ms) # The time managed threads were paused during this GC, in milliseconds
- reason # Reason for GC.
- suspension time (ms) # The time in milliseconds that it took to suspend all threads to start this GC 
- pause time (%) # The amount of time that execution in managed code is blocked because the GC needs exclusive use to the heap. For background GCs this is small.
- gen0 alloc (mb) # Amount allocated in Gen0 since the last GC occurred in MB.
- gen0 alloc rate #  The average allocation rate since the last GC.
- peak size (mb) # The size on entry of this GC (includes fragmentation) in MB.
- after size (mb) # The size on exit of this GC (includes fragmentation) in MB.
- peak/after # Peak / After
- promoted (mb) # Memory this GC promoted in MB.
- finalize promoted (mb) # The size of finalizable objects that were discovered to be dead and so promoted during this GC, in MB.
- pinned objects # Number of pinned objects observed in this GC.
- gen0 size (mb) # Size of gen0 at the end of this GC in MB.
- gen0 survival rate # The % of objects in gen0 that survived this GC.
- gen0 frag ratio # The % of fragmentation on gen0 at the end of this GC.
- gen1 size (mb) # Size of gen1 at the end of this GC in MB.
- gen1 survival rate # The % of objects in gen1 that survived this GC.
- gen1 frag ratio # The % of fragmentation on gen1 at the end of this GC.
- gen2 size (mb) # Size of gen2 at the end of this GC in MB.
- gen2 survival rate # The % of objects in gen2 that survived this GC.
- gen2 frag ratio # The % of fragmentation on gen2 at the end of this GC.
- LOH size (mb) # Size of LOH at the end of this GC in MB.
- LOH survival rate # The % of objects in LOH that survived this GC.
- LOH frag ratio # The % of fragmentation on LOH at the end of this GC.
# Statistic Mode including timer period for printing Heap Stats.
stats_mode: 
# Display conditions such as min gc pause time to display.
display_conditions: 
```

chrisnas commented on src/GCRealTimeMon/Program.cs:
```diff
@@ -21,26 +21,32 @@ public class Options
                     longName: "processName",
                     Required = false,
-                    HelpText = "The process name for which the GC Monitoring will take place for - the first process is chosen if there are multiple.")]
+                    HelpText = "The process name for which the GC Monitoring will take place for - the first process is chosen if there are multiple. Either this parameter or -p must be passed for the monitoring to begin.")]
```
For another PR: remember to change that behavior because you don't want to have monitor an unexpected process.
If more than 1 process with the given name, should exit.
We could also support ps command to list the monitorable processes like what dotnet counters/trace do (see https://github.com/dotnet/diagnostics/blob/7ebf1c10010e1a505914573535b20d8a6f85cc29/src/Tools/Common/Commands/ProcessStatus.cs#L19 for implementation details)

chrisnas commented on src/GCRealTimeMon/Program.cs:
```diff
@@ -21,26 +21,32 @@ public class Options
                     longName: "configPath",
                     Required = false,
-                    HelpText = "The path to the YAML configuration file that is read in.")]
+                    HelpText = "The path to the YAML configuration file that is used in the session; this configuration file contains information such as columns to display that can be configured accordingly. If no path is specified, the default configuration is overwritten by the answers to the prompts in the command line.")]
```
```suggestion
HelpText = "The path to the YAML columns configuration file used during the session. If no path is specified, the default configuration is overwritten by the selected column in the prompt.")]
```

chrisnas commented on src/GCRealTimeMon/Program.cs:
```diff
@@ -21,26 +21,32 @@ public class Options
+            [Option(shortName: 'g',
+                    longName: "createConfigPath",
+                    Required = false,
+                    HelpText = "The path of the config to be created via the command line prompt.")]
```
```suggestion
                    HelpText = "The path of the YAML configuration file to be generated based on the colums selection available in the command prompt.")]
```

chrisnas commented on src/GCRealTimeMon/Utilities/CommandLineUtilities.cs:
```diff
@@ -0,0 +1,61 @@
+        public const string RequiredCommandNotProvided = @"
+Required Command was not provided. The monitoring begins only when a process name via -n or a process id via -p is supplied.
+Examples:
+    - GCRealTimeMon.exe -n devenv / dotnet-gcmon -n devenv
```
You need to figure out what is the current executing assembly to make the difference between GCRealTimeMon and dotnet-gcmon (see https://github.com/Maoni0/realmon/blob/main/src/GCRealTimeMon/Program.cs#L183).
This applies to other hardcoded "GCRealTimeMon" in that file

chrisnas commented on src/GCRealTimeMon/Utilities/CommandLineUtilities.cs:
```diff
@@ -0,0 +1,61 @@
+    GCRealTimeMon.exe [command line args] / dotnet-gcmon [command line args]
+
+More Details:
+    - Specify a process Id by using -p or a process name by using -n, the tool will show GCs as they occur in that process. If there are multiple processes with that name it would pick the first one
+    - You can specify which info to display per GC by using a config file. You can either change the current config at ``{Configuration.ConfigurationReader.DefaultPath}`` -- specifically, if you are using this as a dotnet global tool it's the path where gcmon dll is, but if you are using this when you build it yourself, it's the DefaultConfig.yaml where GCRealTimeMon.dll is. Or specify your own b ...
+    - To create a config file, use -g which allows you to specify a path for the config file and choose which info to display for each GC or overwrite the default config by entering -c without any parameters. 
```
```suggestion
    - Specify a process Id with -p or a process name wiht -n to display details of the garbage collection as they occur in that process. If there are multiple processes with that name it would pick the first one.
    - Displayed columns per GC are defined in a YAML configuration file. It can be changed to a custom one generated with -g or manually select columns with -c. 
```

MokoSan committed:
Updated the "-c" description.

Updated the configuration details with a better description.

Co-authored-by: chrisnas <chrisnas@users.noreply.github.com>

MokoSan committed:
Updated the "-g" description

Updated the -g configuration details with a better description.

Co-authored-by: chrisnas <chrisnas@users.noreply.github.com>

MokoSan committed:
Incorporated feedback

MokoSan committed:
Updated the Help Text for the -g option

Updated the Help Text for the -g option

Co-authored-by: chrisnas <chrisnas@users.noreply.github.com>

MokoSan committed:
Updated the Help Text for the -c option

Updated the Help Text for the -c option

Co-authored-by: chrisnas <chrisnas@users.noreply.github.com>

MokoSan committed:
Updated the `More Details` text

MokoSan committed:
Merge branch 'dev/musharm/ConfigCreator' of https://github.com/MokoSan/realmon into dev/musharm/ConfigCreator

MokoSan committed:
Added the full default path

MokoSan commented:
@chrisnas @Maoni0, I think this PR is ready to be merged. Please let me know if you think any other changes are needed / have any other feedback. 

Thanks for all your help! Been a great learning experience.

Maoni0 reviewed (approved):
LGTM! I tried it a little and it seems to work fine. @chrisnas do you have any comments?

chrisnas commented:
> LGTM! I tried it a little and it seems to work fine. @chrisnas do you have any comments?

If all my comments were taken into account, goooooooooooooo! 
P.S. great PR  :^)

Maoni0 commented:
thanks sooo much for doing this! 🙂

Maoni0 commented:
I'll publish a new version to nuget tonight

MokoSan commented:
![image](https://user-images.githubusercontent.com/4951960/143089825-977a95d3-d23f-453b-adcb-50567f636c55.png)

I double checked the changes and I think we are good. I am going to do some additional testing once the new nuget package is published.

Thanks again for your help, @Maoni0 and @chrisnas!
